eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -146,421 +129,143 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: A
+      d.channels.get(FinalChannelId(msg.channelId)) match {
+        case Some(channel) => channel forward msg
+        case None => d.peerConnection ! wire.Error(msg.channelId, UNKNOWN_CHANNEL_MESSAGE)
+      }
+      stay
 
-      case Event(ResumeAnnouncements, d: ConnectedData) =>
-        log.info(s"resuming processing of network announcements for peer")
-        stay using d.copy(behavior = d.behavior.copy(fundingTxAlreadySpentCount = 0, ignoreNetworkAnnouncement = false))
+    case Event(msg: wire.HasTemporaryChannelId, d: ConnectedData) =>
+      d.channels.get(TemporaryChannelId(msg.temporaryChannelId)) match {
+        case Some(channel) => channel forward msg
+        case None => d.peerConnection ! wire.Error(msg.temporaryChannelId, UNKNOWN_CHANNEL_MESSAGE)
+      }
+      stay
 
-      case Event(Disconnect(nodeId), d: ConnectedData) if nodeId == remoteNodeId =>
-        log.info(s"disconnecting")
-        sender ! "disconnecting"
-        d.transport ! PoisonPill
-        stay
+    case Event(ChannelIdAssigned(channel, _, temporaryChannelId, channelId), d: ConnectedData) if d.channels.contains(TemporaryChannelId(temporaryChannelId)) =>
+      log.info(s"channel id switch: previousId=$temporaryChannelId nextId=$channelId")
+      // NB: we keep the temporary channel id because the switch is not always acknowledged at this point (see https://github.com/lightningnetwork/lightning-rfc/pull/151)
+      // we won't clean it up, but we won't remember the temporary id on channel termination
+      stay using d.copy(channels = d.channels + (FinalChannelId(channelId) -> channel))
 
-      case Event(Terminated(actor), d: ConnectedData) if actor == d.transport =>
-        Logs.withMdc(diagLog)(Logs.mdc(category_opt = Some(Logs.LogCategory.CONNECTION))) {
-          log.info(s"lost connection to $remoteNodeId")
-        }
-        if (d.channels.isEmpty) {
-          // we have no existing channels, we can forget about this peer
-          stopPeer()
-        } else {
-          d.channels.values.toSet[ActorRef].foreach(_ ! INPUT_DISCONNECTED) // we deduplicate with toSet because there might be two entries per channel (tmp id and final id)
-          goto(DISCONNECTED) using DisconnectedData(d.address_opt, d.channels.collect { case (k: FinalChannelId, v) => (k, v) })
-        }
-
-      case Event(Terminated(actor), d: ConnectedData) if d.channels.values.toSet.contains(actor) =>
-        // we will have at most 2 ids: a TemporaryChannelId and a FinalChannelId
-        val channelIds = d.channels.filter(_._2 == actor).keys
-        log.info(s"channel closed: channelId=${channelIds.mkString("/")}")
-        if (d.channels.values.toSet - actor == Set.empty) {
-          log.info(s"that was the last open channel, closing the connection")
-          d.transport ! PoisonPill
-        }
-        stay using d.copy(channels = d.channels -- channelIds)
+    case Event(Disconnect(nodeId), d: ConnectedData) if nodeId == remoteNodeId =>
```

Comment:
It looks like if we manually disconnect via this event, the reconnectionTask will still reconnect us afterwards :confounded: 
Maybe we need a field in `DisconnectedData` to mention that it's disconnected from a user request and should stay disconnected until a `Peer.Connect`?

Reply:
> It looks like if we manually disconnect via this event, the reconnectionTask will still reconnect us afterwards

Yes, but that will happen only if we have existing channels and I don't know what the use case would be for not reconnecting in that case. FWIW `Disconnect` is mostly useful for tests and debugging, and we can also disable automated reconnection altogether with the `auto-reconnect` config parameter (which we use in tests).

Reply:
That's true, it's probably an edge-case we can safely ignore